### PR TITLE
kvserver: run LeaseFromExpirationToEpochDoesNotRegressExpiration with leader leases

### PR DESCRIPTION
**Pull request by iskettaneh**
This commit allows the test
TestLeaseRequestFromExpirationToEpochDoesNotRegressExpiration to run with both epoch-base leases and leader leases.

References: #133763

Release note: None

### CODE REVIEW

**Review by arulajmani — COMMENTED** [review]
> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136267)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`pkg/kv/kvserver/client_lease_test.go` line 1702 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9DLp2TsIp7aTkEtI:-OCf9DLp2TsIp7aTkEtJ:bcr5zzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1702)):*
> ```Go
> 				ts = l.Expiration.ToTimestamp()
> 			} else if leaseType == roachpb.LeaseLeader {
> 				// Make sure that store liveness heartbeat has started.
> ```

It may be worth extracting this as a function on `TestCluster` -- `tc.EnsureStoreLivenessHeartbeatsToSelfHaveStarted`.

Also, it's not obvious to me why we need this -- could you say more (and maybe even capture that in the comment).
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf8jy5-AGLy1bK5tTN:-OCf8jy5-AGLy1bK5tTO:bjq8vrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1703)):*
> ```Go
> 			} else if leaseType == roachpb.LeaseLeader {
> 				// Make sure that store liveness heartbeat has started.
> 				store, err := tc.Servers[0].GetStores().(*kvserver.Stores).
> ```

Should we use `GetFirstStoreFromServer` here instead?
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1717 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9j3V0OPYWJL7VTDJ:-OCf9j3V0OPYWJL7VTDK:b-gyb33q) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1717)):*
> ```Go
> 
> 				// Stop store liveness heartbeats.
> 				var partitioned atomic.Bool
> ```

Could we supply nil instead to `dropStoreLivenessHeartbeatsFrom`?
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1749 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9rh86gT9p-0ikjp9:-OCf9rh86gT9p-0ikjpA:b-9p733l) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1749)):*
> ```Go
> 			kvserver.OverrideDefaultLeaseType(ctx, &s0.ClusterSettings().SV, leaseType)
> 
> 			// Wait for that lease promotion to occur.
> ```

Should we use `WaitForLeaseUpgrade` here instead?
If we do, we'll want to make sure that the lease returned is equal to the version of the test we're running.
___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf980X0fVal-nx7cOq:-OCf980X0fVal-nx7cOr:bwgqg8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_raft_helpers_test.go#L462)):*
> ```Go
> // that drops inbound store liveness messages from the given range and replica's
> // store.
> func dropStoreLivenessHeartbeatsFrom(
> ```

Should we add some words, like above, about what happens when `cond == nil`?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136267)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/client_lease_test.go` line 1702 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9DLp2TsIp7aTkEtI:-OCiB49h2jLvfHF-iStO:bqsxbsd) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1702)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

It may be worth extracting this as a function on `TestCluster` -- `tc.EnsureStoreLivenessHeartbeatsToSelfHaveStarted`.

Also, it's not obvious to me why we need this -- could you say more (and maybe even capture that in the comment).
</blockquote></details>

Just because the timestamp starts at 0. I wasn't sure if an empty timestamp might be treated differently or not. I thought that's it's good that we stop store liveness heartbeats only after they start.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1703 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf8jy5-AGLy1bK5tTN:-OCi5gayEZ4Qnep7b0nb:b6sq8r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1703)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we use `GetFirstStoreFromServer` here instead?
</blockquote></details>

We use the actual store object below
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1717 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9j3V0OPYWJL7VTDJ:-OCi5PlJ4HKdjZk4oaEd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1717)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could we supply nil instead to `dropStoreLivenessHeartbeatsFrom`?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1749 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9rh86gT9p-0ikjp9:-OCiPRVqDXX75j9NZcHg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1749)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we use `WaitForLeaseUpgrade` here instead?
If we do, we'll want to make sure that the lease returned is equal to the version of the test we're running.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_raft_helpers_test.go` line 462 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf980X0fVal-nx7cOq:-OCi5HAUDLDWkhNx3CjI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_raft_helpers_test.go#L462)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we add some words, like above, about what happens when `cond == nil`?
</blockquote></details>

Done.

**Review by arulajmani — COMMENTED** [review]
> Reviewed 13 of 13 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136267)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh, @jbowens, and @kev-cao)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjOsYu1a5gYw9l2wmz:-OCjOsYu1a5gYw9l2wn-:bs867v8):*
I don't think this was intentional, was it? We should drop this commit before merging. 
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1702 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9DLp2TsIp7aTkEtI:-OCjR0xk0x84SMc6ecwl:bmd32v8) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1702)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Just because the timestamp starts at 0. I wasn't sure if an empty timestamp might be treated differently or not. I thought that's it's good that we stop store liveness heartbeats only after they start.
</blockquote></details>

We spoke about this offline and decided to get rid of it.
___
*[`pkg/testutils/testcluster/testcluster.go` line 12 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjP7l06_6cMencxV0g:-OCjP7l06_6cMencxV0h:bc95nvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1918904407b9b1dad265e96e8115e1282d2c484/pkg/testutils/testcluster/testcluster.go#L12)):*
> ```Go
> 	gosql "database/sql"
> 	"fmt"
> 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
> ```

Is crlfmt happy about this?
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1577 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjPJ6yEEKOfd6FPwNk:-OCjPJ6yEEKOfd6FPwNl:b-tqrn28) ([raw file](https://github.com/cockroachdb/cockroach/blob/1451eb3d76805291d7a81f398a9b90834bbc70eb/pkg/kv/kvserver/client_lease_test.go#L1577)):*
> ```Go
> 
> 	// Expect it to be upgraded to an epoch based lease.
> 	epochL, _ := tc.WaitForLeaseUpgrade(ctx, t, desc)
> ```

Should this have been part of the previous commit?

**Review by iskettaneh — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136267)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @jbowens, and @kev-cao)

___
*[`-- commits` line 2 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjOsYu1a5gYw9l2wmz:-OCkNzZ--a_bv0u8B_fE:b-896fix):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I don't think this was intentional, was it? We should drop this commit before merging.
</blockquote></details>

Done.
___
*[`pkg/testutils/testcluster/testcluster.go` line 12 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjP7l06_6cMencxV0g:-OCkPep69SV1uvtJcXWM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e1918904407b9b1dad265e96e8115e1282d2c484/pkg/testutils/testcluster/testcluster.go#L12)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Is crlfmt happy about this?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1702 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCf9DLp2TsIp7aTkEtI:-OCkOWWj8q-7iRE1pLAJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/af8f0b62573e454777dae073b97f9f67b1e5305b/pkg/kv/kvserver/client_lease_test.go#L1702)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

We spoke about this offline and decided to get rid of it.
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1577 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjPJ6yEEKOfd6FPwNk:-OCkPbnf8ClTbU3Zjf0l:buydgb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1451eb3d76805291d7a81f398a9b90834bbc70eb/pkg/kv/kvserver/client_lease_test.go#L1577)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should this have been part of the previous commit?
</blockquote></details>

This is the first commit

**Review by arulajmani — APPROVED** [review]
:lgtm: modulo comments

> Reviewed 15 of 15 files at r5, 15 of 15 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-:-OCkZN22B8BVxJFQ4Caj:b-913mr5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @iskettaneh)

___
*[`-- commits` line 6 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCkYaRUBjH3EabP3kxJ:-OCkYaRUBjH3EabP3kxK:b-vnintq):*
" ... etc" ?
___
*[`pkg/kv/kvserver/client_lease_test.go` line 1577 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCjPJ6yEEKOfd6FPwNk:-OCkZHDo3qqnSaSNksYj:b-b2i1eh) ([raw file](https://github.com/cockroachdb/cockroach/blob/1451eb3d76805291d7a81f398a9b90834bbc70eb/pkg/kv/kvserver/client_lease_test.go#L1577)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

This is the first commit
</blockquote></details>

I think reviewable got confused because the PR went from 2 commits to 3. Sorry about that.
___
*[`pkg/testutils/testcluster/testcluster.go` line 1594 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/136267#-OCkYmnf9OkKGEJR5za_:-OCkYmnf9OkKGEJR5zaa:bzgnmi9) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d638e381263a5145cbb1d900aa4f8c000639bd0/pkg/testutils/testcluster/testcluster.go#L1594)):*
> ```Go
> // liveness heartbeat to self has started (and not empty). Returns the timestamp
> // of the latest supportFrom() from itself.
> func (tc *TestCluster) EnsureStoreLivenessHeartbeatsToSelfHaveStarted(
> ```

I think we can get rid of this now that it's no longer used.

**Comment by iskettaneh** [comment]
TFTR!

bors r+

**Pull request merged**
